src/cuter_types.erl

```diff
@@ -804,15 +804,124 @@ maybe_hash_args(Ps, Args) ->
   Ps ++ [integer_to_list(Arity), integer_to_list(erlang:phash2(Args))].
 
 %% ----------------------------------------------------------------------------
+%% Normalize parsed specs & type dependencies
+%% ----------------------------------------------------------------------------
+
+normalize_type_deps(Deps) ->
+  Seen = ets:new(?MODULE, [ordered_set, protected]),
+  NormalizedDeps = normalize_type_deps(Deps, Seen, []),
+  ets:delete(Seen),
+  NormalizedDeps.
+
+normalize_type_deps([], _Seen, Acc) ->
+  lists:reverse(Acc);
+normalize_type_deps([{TypeName, Type}|Types], Seen, Acc) ->
+  {NormalizedType, MoreDeps} = normalize_single_type(Type, Seen, true),
+  Acc1 = lists:reverse(MoreDeps) ++ Acc,
+  normalize_type_deps(Types, Seen, [{TypeName, NormalizedType}|Acc1]).
+
+normalize_single_type(#t{kind = ?function_tag, rep = {Params, Ret, _}}=Type, Seen, IsTopLevel) ->
+  case ets:lookup(Seen, Type) of
+    [{Type, Handle}] ->
+      {Handle, []};
+    [] ->
+      Xs = [normalize_single_type(T, Seen, false) || T <- [Ret|Params]],
+      {Ts, Deps} = lists:unzip(Xs),
+      NormalizedType = Type#t{rep = {tl(Ts), hd(Ts), []}},
+      AllDeps = lists:flatten(lists:map(fun lists:reverse/1, Deps)),
+      case IsTopLevel of
+        true ->
+          {NormalizedType, AllDeps};
+        false ->
+          NewTypeName = generate_new_type(),
```

Comment:
I see the same five lines being repeated a number of times below.  They are a prime candidate to be taken out into a function and avoid the code duplication...